qiskit_aer/quantum_info/states/aer_statevector.py

```diff
@@ -0,0 +1,273 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Statevector quantum state class.
+"""
+import copy
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit, Instruction
+from qiskit.quantum_info.states import Statevector
+from qiskit.quantum_info.operators.operator import Operator
+
+from qiskit_aer import AerSimulator
+from .aer_state import AerState
+from ...backends.aerbackend import AerError
+
+
+class AerStatevector(Statevector):
+    """AerStatevector class"""
+
+    def __init__(self, data, **configs):
+        """Initialize a statevector object.
+
+        Args:
+            data (np.array or list or Statevector or QuantumCircuit or qiskit.circuit.Instruction):
+                Data from which the statevector can be constructed. This can be either a complex
+                vector, another statevector or a ``QuantumCircuit`` or ``Instruction``
+                (``Operator`` is not supportted in the current implementation).  If the data is
+                a circuit or instruction, the statevector is constructed by assuming that all
+                qubits are initialized to the zero state.
+            configs (kwargs): configurations of ``AerState``. All the keys are handeld as string
+
+        Raises:
+            AerError: if input data is not valid.
+
+        Additional Information:
+            The ``dims`` kwarg is used to ``Statevector`` constructor.
+
+        """
+        if 'dims' in configs:
+            dims = configs.pop('dims')
+        else:
+            dims = None
+
+        if isinstance(data, (QuantumCircuit, Instruction)):
+            data, aer_state = AerStatevector._from_instruction(data, None, **configs)
+        elif isinstance(data, list):
+            data, aer_state = AerStatevector._from_ndarray(np.array(data, dtype=complex), **configs)
+        elif isinstance(data, np.ndarray):
+            data, aer_state = AerStatevector._from_ndarray(data, **configs)
+        elif isinstance(data, AerStatevector):
+            aer_state = data._aer_state
+            if dims is None:
+                dims = data._op_shape._dims_l
+            data = data._data
+        else:
+            raise AerError(f'Input data is not supported: type={data.__class__}, data={data}')
+
+        super().__init__(data, dims=dims)
+        self._aer_state = aer_state
+        self._result = None
+        self._configs = configs
+
+    def _last_result(self):
+        if self._result is None:
+            self._result = self._aer_state.last_result()
+        return self._result
+
+    @property
+    def metadata(self):
+        """Return result metadata of an operation that executed lastly."""
+        if self._last_result() is None:
+            raise AerError('AerState was not used and metdata does not exist.')
+        return self._last_result()['metadata']
```

Comment:
Unless this property is meant to be mutable, it would be clearer to make it a function, not a property, since it's often recalculated, and access can raise an exception.